### PR TITLE
[SPARK-3688][SQL]LogicalPlan can't resolve column correctlly

### PR DESCRIPTION
**Pull request by tianyi**
This PR fixed the resolving problem described in https://issues.apache.org/jira/browse/SPARK-3688

```
CREATE TABLE t1(x INT);
CREATE TABLE t2(a STRUCT<x: INT>, k INT);
SELECT a.x FROM t1 a JOIN t2 b ON a.x = b.k;
```

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -128,6 +128,29 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
   def resolve(name: String, resolver: Resolver): Option[NamedExpression] =
     resolve(name, output, resolver)
 
+  def resolveAsTableColumn(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: If we are going to merge this, can you add javadoc for these two methods to explain what they do, and what the differences are?

**Comment by marmbrus** [comment]
I merged this since we're close to the deadline.  Thanks!

Please address @rxin comments in a follow up.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 1 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Also they should be private or at least protected.

**Comment by rxin** [comment]
FYI I added more inline comment https://github.com/apache/spark/pull/4539

**Pull request closed without merging**